### PR TITLE
Add support for downloading folder-based agents

### PR DESCRIPTION
**Pull request by robin-checkmk**
## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Folder-specific agent configurations were not considered by the agent role.

## What is the new behavior?

The agent role now tries to download the host-specific agent, if that fails, it tries the folder-specific agent and falls back to the GENERIC agent, if the former two cannot be found.

### CODE REVIEW

**Comment by robin-checkmk** [comment]
recheck

**Comment by robin-checkmk** [comment]
Postponed until after #156 is merged.

File: roles/agent/tasks/Windows.yml [hunk 1]
```diff
@@ -37,7 +37,11 @@
 - name: "Set Fact: Agent State: host-specific."
   ansible.builtin.set_fact:
     checkmk_agent_host_specific: >-
-     {% if checkmk_agent_host_download_state.status_code is defined and checkmk_agent_host_download_state.status_code == 200 %}true{% else %}false{% endif %}
+     {%- if checkmk_agent_host_download_state.status_code is defined and checkmk_agent_host_download_state.status_code == 200 -%}
```
Comment (msekania) [inline roles/agent/tasks/Windows.yml:40]: @robin-checkmk, just saw it

what's about
{{ true if checkmk_agent_host_download_state.status_code is defined and checkmk_agent_host_download_state.status_code == 200 else false}}

Comment (robin-checkmk) [reply roles/agent/tasks/Windows.yml:40]: Does that syntax work? Never saw something like that.

Comment (msekania) [reply roles/agent/tasks/Windows.yml:40]: https://jinja.palletsprojects.com/en/3.1.x/templates/#if-expression

Comment (msekania) [reply roles/agent/tasks/Windows.yml:40]: In essence in you case its even simpler:
```yaml
{{ (checkmk_agent_host_download_state.status_code is defined and checkmk_agent_host_download_state.status_code == 200) | bool }}
```

Comment (robin-checkmk) [reply roles/agent/tasks/Windows.yml]: Thanks mate, that's way more elegant, love it! :rocket:

File: roles/agent/tasks/Linux.yml [hunk 2]
```diff
@@ -14,17 +14,18 @@
     checkmk_agent_prep_legacy | bool
     and ansible_facts.packages['systemd'][0]['version'] | regex_search('\d{1,}') | int < 220
 
-- name: "{{ ansible_system }}: Download Checkmk CRE agent."
+- name: "{{ ansible_system }}: Download Vanilla {{ checkmk_agent_edition | upper }} agent."
```
Comment (Max-checkmk) [inline roles/agent/tasks/Linux.yml:17]: I would still use the fixed "CRE" within this taskname as this task will only happen if the edition is "cre".

If it is skipped the taskname is as example: "Linux: Download Vanilla CEE agent.", which doesn't make sense.

File: roles/agent/tasks/Windows.yml [hunk 3]
```diff
@@ -1,18 +1,18 @@
 ---
-- name: "{{ ansible_os_family }}: Download Checkmk CRE Agent."
+- name: "{{ ansible_os_family }}: Download Vanilla {{ checkmk_agent_edition | upper }} Agent."
```
Comment (Max-checkmk) [inline roles/agent/tasks/Windows.yml:2]: I would still use the fixed "CRE" within this taskname as this task will only happen if the edition is "cre".

If it is skipped the taskname is as example: "Linux: Download Vanilla CEE agent.", which doesn't make sense.

File: roles/agent/tasks/Windows.yml [hunk 4]
```diff
@@ -57,33 +82,27 @@
...
-    checkmk_agent_edition == "cfe" or
-    checkmk_agent_edition == "cme")
-    and checkmk_agent_download_state.status_code != 200
+    checkmk_agent_edition | lower != "cre"
+    and not checkmk_agent_host_specific | bool
   tags:
     - install-package
 
-- name: "{{ ansible_os_family }}: Install Checkmk CRE Agent."
+- name: "{{ ansible_os_family }}: Install Vanilla {{ checkmk_agent_edition | upper }} Agent."
```
Comment (Max-checkmk) [inline roles/agent/tasks/Windows.yml:102]: I would still use the fixed "CRE" within this taskname as this task will only happen if the edition is "cre".

If it is skipped the taskname is as example: "Linux: Install Vanilla CEE agent.", which doesn't make sense.

File: roles/agent/tasks/Debian.yml [hunk 5]
```diff
@@ -63,52 +86,54 @@
...
-    checkmk_agent_edition == "cfe" or
-    checkmk_agent_edition == "cme")
-    and checkmk_agent_download_state.status != 200
+    checkmk_agent_edition | lower != "cre"
+    and not checkmk_agent_host_specific | bool
   tags:
     - install-package
 
-- name: "{{ ansible_os_family }} Derivatives: Install Checkmk CRE Agent."
+- name: "{{ ansible_os_family }} Derivatives: Transfer Vanilla {{ checkmk_agent_edition | upper }} Agent."
```
Comment (Max-checkmk) [inline roles/agent/tasks/Debian.yml:120]: I would still use the fixed "CRE" within this taskname as this task will only happen if the edition is "cre".

If it is skipped the taskname is as example: "...Transfer Vanilla CEE agent.", which doesn't make sense.

File: roles/agent/tasks/RedHat.yml [hunk 6]
```diff
@@ -63,53 +86,55 @@
...
-    checkmk_agent_edition == "cfe" or
-    checkmk_agent_edition == "cme")
-    and checkmk_agent_download_state.status != 200
+    checkmk_agent_edition | lower != "cre"
+    and not checkmk_agent_host_specific | bool
   tags:
     - install-package
 
-- name: "{{ ansible_os_family }} Derivatives: Install Checkmk CRE Agent."
+- name: "{{ ansible_os_family }} Derivatives: Transfer Vanilla {{ checkmk_agent_edition | upper }} Agent."
```
Comment (Max-checkmk) [inline roles/agent/tasks/RedHat.yml:120]: I would still use the fixed "CRE" within this taskname as this task will only happen if the edition is "cre".

If it is skipped the taskname is as example: "...Transfer Vanilla CEE agent.", which doesn't make sense.

File: roles/agent/tasks/Suse.yml [hunk 7]
```diff
@@ -64,54 +87,56 @@
...
-    checkmk_agent_edition == "cfe" or
-    checkmk_agent_edition == "cme")
-    and checkmk_agent_download_state.status != 200
+    checkmk_agent_edition | lower != "cre"
+    and not checkmk_agent_host_specific | bool
   tags:
     - install-package
 
-- name: "{{ ansible_os_family }} Derivatives: Install Checkmk CRE Agent."
+- name: "{{ ansible_os_family }} Derivatives: Transfer Vanilla {{ checkmk_agent_edition | upper }} Agent."
```
Comment (Max-checkmk) [inline roles/agent/tasks/Suse.yml:122]: I would still use the fixed "CRE" within this taskname as this task will only happen if the edition is "cre".

If it is skipped the taskname is as example: "... Transfer Vanilla CEE agent.", which doesn't make sense.

Comment (robin-checkmk) [reply roles/agent/tasks/Suse.yml]: Fixed it. Go away now.

File: roles/agent/tasks/RedHat.yml (hunk 6 above)
Comment (robin-checkmk) [reply roles/agent/tasks/RedHat.yml]: Fixed it. Go away now.

File: roles/agent/tasks/Debian.yml (hunk 5 above)
Comment (robin-checkmk) [reply roles/agent/tasks/Debian.yml]: Fixed it. Go away now.

File: roles/agent/tasks/Windows.yml (hunk 3 above)
Comment (robin-checkmk) [reply roles/agent/tasks/Windows.yml]: Fixed it. Go away now.

**Review by Max-checkmk — APPROVED** [review]
Works and looks fine for me! ;)

**Pull request merged**